pkg/typesystem/connected_types.go

```diff
@@ -0,0 +1,123 @@
+package typesystem
+
+import (
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+// map[_objectType_]map[_relation_]map[_subjectType_] terminalRelation.
+type TypesystemConnectedTypes map[string]map[string]map[string][]string
+
+func (f TypesystemConnectedTypes) assign(objectType string, relation string, subjectType string, terminalRelation string) {
+	if f[objectType] == nil {
+		f[objectType] = make(map[string]map[string][]string)
+	}
+
+	if f[objectType][relation] == nil {
+		f[objectType][relation] = make(map[string][]string)
+	}
+
+	f[objectType][relation][subjectType] = append(f[objectType][relation][subjectType], terminalRelation)
+}
+
+// AssignTerminalTypes will populate the `connectedTypes` property on the typesystem to indicate for a given object type
+// what the terminal user types and relations.
+func (t *TypeSystem) AssignTerminalTypes(typeName, relationName string) {
+	terminalTypesAndRelations := t.getTerminalUserTypeAndRelationsForConnectedTypes(typeName, relationName, 0)
+	for _, terminalTypesAndRelation := range terminalTypesAndRelations {
+		for _, terminalType := range terminalTypesAndRelation.terminalTypes {
+			t.connectedTypes.assign(typeName, relationName, terminalType, terminalTypesAndRelation.terminalRelation)
+		}
+	}
+}
+
+type terminalTypesAndRelation struct {
+	terminalTypes    []string
+	terminalRelation string
+}
+
+func (t *TypeSystem) getTerminalUserTypeAndRelationsForConnectedTypes(
+	typeName string,
+	relationName string,
+	numTTU int,
+) []terminalTypesAndRelation {
+	rewrite := t.typeDefinitions[typeName].GetRelations()[relationName]
+
+	cache, ok := t.connectedTypes[typeName][relationName]
+	if ok {
+		terminalTypesAndRelations := []terminalTypesAndRelation{}
+
+		for subjectType, terminalRelations := range cache {
+			for _, terminalRelation := range terminalRelations {
+				terminalTypesAndRelations = append(terminalTypesAndRelations, terminalTypesAndRelation{
+					terminalTypes:    []string{subjectType},
+					terminalRelation: terminalRelation,
+				})
+			}
+		}
+
+		return terminalTypesAndRelations
+	}
+
+	relation, ok := t.relations[typeName][relationName]
```

Comment:
question: Do we need to check for `t.relations[typeName][relationName]` in the case rewrite.GetUserset().(type) is not `*openfgav1.Userset_This`?  I see that we only use relation in the "this" path.  (Or this is a case where we should never happen)..

Reply:
Good point. I've change this to only apply to the `Userset_This` rewrite.